Assets/Engine/Audio/AudioManager.cs

```diff
@@ -0,0 +1,204 @@
+        /// </summary>
+        public void PlayAudioSource(AudioClip audioClip, GameObject parent)
+        {
+            PlayAudioSource(audioClip, parent.transform.position, parent);
+        }
+    
+        /// <summary>
+        /// Plays a sound clip at a position, parent, with specific volume, pitch, and ranges.
+        /// Volume, pitch, and ranges are optional.
+        /// </summary>
+        public void PlayAudioSource(AudioClip audioClip, Vector3 position, GameObject parent, float volume = 0.7f, float pitch = 1f, float minRange = 1f, float maxRange = 500f)
+        {
+            var audioSource = FindAvailableAudioSource();
+            audioSource.gameObject.transform.position = position;
+            audioSource.clip = audioClip;
+            audioSource.volume = volume;
+            audioSource.pitch = pitch;
+            audioSource.minDistance = minRange;
+            audioSource.maxDistance = maxRange;
+            //If we want to attach the audio source to something specific, do that. Otherwise, detach it from any parents.
+            //This is useful for things that are obviously creating the sound, like a mouse's squeak -- we don't want the mouse to leave the squeak behind as it travels, but a flying soda can making a sound at the site of impact is probably fine.
+            audioSource.transform.parent = parent == null ? null : parent.transform;
+            audioSource.Play();
+        }
+
+        /// <summary>
+        /// Plays a sound clip using a specific audio source at a position, parent, with specific volume, pitch, and ranges.
+        /// Volume, pitch, and ranges are optional. 
+        /// </summary>
+        public void PlayAudioSourceSpecific(AudioSource audioSource, AudioClip audioClip, Vector3 position, GameObject parent, float volume = 0.7f, float pitch = 1f, float minRange = 1f, float maxRange = 500f)
+        {
+            audioSource.gameObject.transform.position = position;
+            audioSource.clip = audioClip;
+            audioSource.volume = volume;
+            audioSource.pitch = pitch;
+            audioSource.minDistance = minRange;
+            audioSource.maxDistance = maxRange;
+            //If we want to attach the audio source to something specific, do that. Otherwise, detach it from any parents.
+            //This is useful for things that are obviously creating the sound, like a mouse's squeak -- we don't want the mouse to leave the squeak behind as it travels, but a flying soda can making a sound at the site of impact is probably fine.
+            audioSource.transform.parent = parent == null ? null : parent.transform;
+            audioSource.Play();            
+        }
+
+        /// <summary>
+        /// Finds an available audio source in the list of available audio sources. If one does not exist, it creates one.
+        /// </summary>
+        public AudioSource FindAvailableAudioSource()
```

Comment:
How about we change this, we depend on `isPlaying` being false here to determine if an audio source is available for not, but what if the script which uses the audio source has a moment where it may not be playing (i.e. delayed sounds, multiple sounds)

Instead we should wrap the audio source in basic container like `RequestedAudioSource` which looks like:

```csharp
// Use these in the AudioManager instead of just AudioSources directly, gives us more control.
// When we create a new audio source we wrap it with this object to act as a reference.
public class WrappedAudioSource
{
   public AudioSource Source { get; private set; }
   public bool Used { get; private set; }

   public WrappedAudioSource(AudioSource source)
   {
       Source = source;
   }

   public void Release()
   {
       // Mark source as free so we know we can use it.
       // Reset transform and source object to its starting state?
       Used = false;
   }
}
```

Reply:
> How about we change this, we depend on `isPlaying` being false here to determine if an audio source is available for not, but what if the script which uses the audio source has a moment where it may not be playing (i.e. delayed sounds, multiple sounds)

Right now audio sources _should_ only be called the moment they are intended to be used, and can be taken by others when they are unused. This was an issue at first with the bike horn, because the bike horn checks if the audio source is still playing in order to decide whether it can honk again. I fixed this by having the bike horn check if the audio source it used is still a child of the bike horn in addition to still making noise. If not, it just takes an unused one. I was trying to avoid having scripts on the actual audio sources as much as possible for performance reasons, though I'm not sure what the actual cost would be like. John suggested that rather than use coroutines to check if they're done playing, we should implement such functionality with UniTask.

That said, I'm not saying no, just wondering if it's necessary.

Reply:
Gonna merge this PR. If this becomes an issue later I don't think it would be terribly difficult to change then.